content/drafts/lib-is-for-libraries.md

```diff
@@ -0,0 +1,113 @@
+
+## So, what _is_ `lib` for?
+
+In a [2012 post](https://codeclimate.com/blog/what-code-goes-in-the-lib-directory/) on the Code Climate blog, I found this statement:
+
+> I recommend any code that is not specific to the domain of the application goes in `lib/`.
+
+🤔 Now this is quite different. The heart of your application would still live in `app/`, but code that is _not specific to its domain_ would go into `lib/`.
+
+Thinking about code that an application uses that's not specific to its domain, examples that come up for me are HTTP clients, authentication and authorization logic, serialization and deserialization, and so on. For these concerns, we usually don't write the code ourselves - we use libraries, gems. Almost always those are third party gems from [rubygems.org](https://rubygems.org/).
+
+But what if what we need is not available in a well maintained library? What if the functionality we need is so small that it doesn't warrant introducing an external dependency? Something that we can easily write ourselves?
+
+We want the code to live in the same repository as our application, but it's not specific to our application's domain. It's still library code. So it should go into `lib/`.
+
+_It's right there in the name!_
+
+**`lib/` is for libraries.**
+
+## Boundaries
+
+The main advantage of separating library code from your application instead of just lumping it in with all of your other stuff is the reduction of cognitive load. You can look at the library code and understand it without having to know any of the details of your application.
+
+This, more generally, is the advantage we get from having strong architectural boundaries. Robert Martin defines boundaries as lines in an architecture diagram that are only crossed by dependencies in one direction. The application depends on libraries, but libraries never depend on the application.
+
+Sandi Metz, in her book _Practical Object-Oriented Design in Ruby_, explains why this matters:
+
+> "I know who you are and because of that I know what you do." This knowledge is a dependency that raises the cost of change.
+
+Dependencies are _knowledge_. And the less you as a developer need to know to do your work the faster you can get things done, delivering value to your users and checking items off your to-do list.
+
+## Complication: Autoloading
+
+In Rails versions before 7.1, `lib/` is not autoloaded, which means that if you want to use code from `lib/` in your application, you need to explicitly require it. While in theory this is a good thing because it makes a dependency explicit, requiring a file will add its contents to the global namespace. That means it will be available everywhere, and implicit dependencies on it will creep in. So the advantage of explicit dependencies doesn't really exist in practice. Also, the code will not automatically be reloaded after making changes locally, which is inconvenient.
```

Comment:
I don't want to muddy the waters of this paper, but this section reminded me of an idea that doesn't get expressed very often - I offer it to you if you want to use it, but feel free to ignore it too (especially if the focus of this post is for a different, more senior audience). The idea is this:

 **if you learned to program with Rails (or most other frameworks), the idea of package boundaries might not be intuitive.** Rails "helps" by making everything available everywhere, to the point that a lot of code-arranging suggestions from the early days ("Fat models, thin controllers", etc) are just conventions. Learning to draw boundaries for code you write yourself (especially within a framework context) is a separate skill, and tooling like packwerk can help.

Reply:
I have expressed this at times :D I'll see if I can fit it in here. I guess I've been hinting at it anyway.

Reply:
I wonder, actually, if "most other frameworks" hwere really means "most other frameworks modeled after rails", since frameworks encouraging strong boundaries do exist (e.g. dry-system).